src/main/java/frc/robot/subsystems/RotatorSubsystem.java

```diff
@@ -21,98 +18,131 @@
         rotationMotor.setInverted(true);
 
         pid = new PIDController(
-                Constants.RotatorConstants.ROTATOR_P_COEFF,
-                Constants.RotatorConstants.ROTATOR_I_COEFF,
-                Constants.RotatorConstants.ROTATOR_D_COEFF);
+                ROTATOR_P_COEFF,
+                ROTATOR_I_COEFF,
+                ROTATOR_D_COEFF);
         // pid.enableContinuousInput(0, Math.PI * 2);
         pid.setTolerance(Math.toRadians(30));
         feedforward = new ArmFeedforward(
-                Constants.RotatorConstants.ROTATOR_KS_COEFF,
-                Constants.RotatorConstants.ROTATOR_KG_COEFF,
-                Constants.RotatorConstants.ROTATOR_KV_COEFF,
-                Constants.RotatorConstants.ROTATOR_KA_COEFF);
+                ROTATOR_KS_COEFF,
+                ROTATOR_KG_COEFF,
+                ROTATOR_KV_COEFF,
+                ROTATOR_KA_COEFF);
 
         encoder = new DutyCycleEncoder(0);
+        // update the rotator PID to the current position
         setToCurrentPosition();
     }
 
+    // Sets the setpiint to where the rotator is currently
     public void setToCurrentPosition() {
         setSetpoint(Math.toDegrees(getMeasurement()));
     }
-    
 
+    // Returns the position of the rotator in radians
     private double getMeasurement() {
         double currentPosition = encoder.getAbsolutePosition();
-        double result = Math.toRadians(((currentPosition * 360) + Constants.RotatorConstants.ENCODER_OFFSET) % 360);
+        double result = Math.toRadians(((currentPosition * 360) + ENCODER_OFFSET) % 360);
         if (log) {
             System.out.println("RotatorPIDOnly.getMeasurement: Current encoder position (raw): " + currentPosition);
             System.out.println(
                     "RotatorPIDOnly.getMeasurement: Current encoder position (adj): " + Math.toDegrees(result));
         }
-    
+
         return result;
     }
 
+    // Returns the position of the rotator in degrees
     public double getMeasurementDegrees() {
         double currentPosition = encoder.getAbsolutePosition();
-        double result = (((currentPosition * 360) + Constants.RotatorConstants.ENCODER_OFFSET) % 360);
-    
+        double result = (((currentPosition * 360) + ENCODER_OFFSET) % 360);
+
         return result;
     }
 
+    // Returns the setpoint of the rotator in degrees
     public double getSetpoint() {
         return Math.toDegrees(pid.getSetpoint());
     }
 
+    // Sets the setpoint of the rotator degrees
     public void setSetpoint(double degrees) {
         pid.setSetpoint(Math.toRadians(degrees));
     }
 
-    //The SetPOINT value is in degrees
-    public void setGoal(double setPoint) {
-        setSetpoint(setPoint);
+    // Sets the goal of the rotator in degrees
+    public void setGoal(double degrees) {
+        setSetpoint(degrees);
     }
 
+    // Sets the speed of the rotator
     public void setSpeed(double speed) {
```

Comment:
This method is only used by the default command for manual control, which works in tandem with the PID. However, as you say, if someone uses this method outside the default command, they will be very sad. 

What's the solution to this? Is it something like the preset power methods we talked about yesterday, or maybe some kind of flag for when this setSpeed is controlling versus the PID in periodic? Is there a way to make this only usable by the default command?

Reply:
That's the thing, I don't see how this even worked with your default command. It's the other spot I commented on, right? Where you set the speed, then immediately tell the PID to track the current position? You said that actually worked at comp, but stop and explain how... on it's face, it shouldn't. Why did it "work"?

So, digging into it, the answer is in the particular implementation of `CommandScheduler.run()`. It runs all subsystems' `periodic`s first, then it runs all currently scheduled commands. So this is what happens:

1. `RotatorSubsystem.perioidic` runs, updates the PID, and sets the motor speed internally.
2. `RotatorDefaultCommand.execute()` runs, sees the stick input, re-updates the motor speed, and then sets the PID to whatever the current position is.
3. 20ms delay before the next loop.
4. `RotatorSubsystem.perioidic` runs, updates the PID which ought to see you're beyond it's setpoint. After all, the rotator moved at manual input speed for 20ms. It then likely computes a new speed of "go backwards towards the setpoint we've passed."
5. `RotatorDefaultCommand.execute()` runs, sees the stick input, re-updates the motor speed, and then sets the PID to whatever the current position is.
6. and that keeps going...

So it "worked" because the default command's action for manual input stomps on the setting periodic just made. And presumably that's fast enough that the motor doesn't seem to stop. Of course, if there are other scheduled commands that need to run before the default command there would be delays between steps 1 & 2, and 4 & 5, and then you might see the jank.

Heh, `m_scheduledCommands` is actually a `Set` backed by a `LinkedHashSet`, so the order in which commands runs is not determined by their scheduling order. It's by their hash in the set. So pushing buttons or running other command could cause delays before or after already scheduled default commands.

So, how to fix?

I see a couple of choices off the top of my head. We can brainstorm more. 

1. Move the `periodic` work into a default command for the rotator. Then you have one place to decide how to control the speed: via the PID, or via manual input. You'd call `setSpeed` when applying manual control, and something like `updateAndFollowPID()` otherwise.
2. Add a "mode" or "state" to the rotator subsystem that says whether the motor is under manual or PID control. Have methods like `startManualMovement(double speed)` and `endManualMovement()` which change the mode, and thus change whether or not you work with the PID in `periodic`.

What's another way? Read this whole page, and specifically this section: https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#instance-command-factory-methods Use a "command factory method" to provide the default command for the subsystem. Define said default command class as a private subclass of the subsystem, and pass the subsystem itself to the constructor of the command. Your default command now has access to the private members of the subsystem. Pick pieces of brain back up off the floor, reassemble and put back in head.